[jit] Enable IValue to hold a PyObject

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30137 [jit] Add RRef to IValue and Jit
* #30630 [rpc] Remvove template on RRef and add Type to RRef creation
* **#30136 [jit] Enable IValue to hold a PyObject**

Summary:

This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D19559730](https://our.internmc.facebook.com/intern/diff/D19559730)

wanchaol committed:
[jit] Enable IValue to hold a PyObject

Summary:

This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Enable IValue to hold a PyObject"


Summary:

This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zdevito commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -60,6 +60,8 @@ TypePtr IValue::type() const {
       return DeviceObjType::get();
     case Tag::Object:
       return toObjectRef().type();
+    case Tag::PythonObject:
```
The tag is PythonObject but the name of the type is PythonObj? Futhermore, python calls these things PyObject and pybind11 its py::object? Can both of these things be renamed to PyObject please?

wanchaol replied:
This is primarily due to name collisions, I can rename `PythonObjType` to `PyObjectType` in jit_type for sure. but in IValue, we need to deal with name collisions between `ivalue::PyObject` and `PyObject` itself.

wanchaol replied:
I think I can keep the tag with `Tag::PyObject`, but with `ivalue::PyObjectHolder` be the real holding class.

zdevito replied:
one is ivalue::Tag::PyObject and the other is ivalue::PyObject, right? Unless I am missing something, it doesn't seem like those conflict.

wanchaol replied:
I did another try on this. the name collision happens between `ivalue::PyObject` and our global python stub `PyObject` defined in https://github.com/pytorch/pytorch/blob/master/c10/util/python_stub.h#L4. C++ compiler messed up with symbol overriding with global namespace.

So the changes I made is that I keep the ivalue class to be `PyObjectHolder`, all other things still be `PyObject`.

zdevito commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -82,6 +82,14 @@ inline c10::intrusive_ptr<ivalue::Object> IValue::toObject() const & {
+  TORCH_INTERNAL_ASSERT(isPythonObject(), "Expected PythonObject but got", tagKind());
+  return moveToIntrusivePtr<ivalue::PythonObject>();
+}
+inline c10::intrusive_ptr<ivalue::PythonObject> IValue::toPythonObject() const & {
```
This should just directly return python `PyObject*` its a pointer so ivalue.h doesn't need to know its definition.

zdevito commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -390,6 +399,21 @@ struct C10_EXPORT ivalue::Object final : c10::intrusive_ptr_target {
+
+  PythonObject(PyObject* py_obj): py_obj_(py_obj) {}
+ private:
+    PyObject* py_obj_;
```
I don't see ref-counting logic for this, meaning this is either a dangling free or a leak of the PyObject.

wanchaol replied:
hmm yeah I was mimicing `ivalue::Object` and I did not see any ref-counting logic for it either... which is weird. I will update with ref-counting logic.

zdevito replied:
This is missing a decref on the PyObject in the destructor of the holder, which will cause this to leak any objects it holds. This will require having PyObjectHolder be abstract in this file and have concrete instances of it be implemented by a subclass in libtorch_python.

zdevito commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -748,6 +772,10 @@ inline IValue::IValue(c10::intrusive_ptr<ivalue::Object> v)
 : tag(Tag::Object), is_intrusive_ptr(true) {
   payload.as_intrusive_ptr = v.release();
 }
+inline IValue::IValue(c10::intrusive_ptr<ivalue::PythonObject> v)
```
We should have a version of the constructor that takes a PyObject or py::object directly and is defined somewhere in libtorch_python.so

zdevito reviewed (commented):
See inline comments for some issues with refcounting and naming.

wanchaol committed:
Update on "[jit] Enable IValue to hold a PyObject"


Summary:

This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Enable IValue to hold a PyObject"


Summary:

This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Enable IValue to hold a PyObject"


Summary:

This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zdevito commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -172,11 +174,16 @@ std::ostream& operator<<(std::ostream & out, const IValue & v) {
-    case IValue::Tag::Object:
+    case IValue::Tag::PyObject: {
+      auto py_obj = v.toPyObjectHolder();
+      return out << "<PyObject at" << py_obj.get() << ">";
```
toPyObject()?

zdevito commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -390,6 +399,30 @@ struct C10_EXPORT ivalue::Object final : c10::intrusive_ptr_target {
+// ivalue Holder that hold a PyObject*
+struct C10_EXPORT ivalue::PyObjectHolder final : c10::intrusive_ptr_target {
+ private:
+  c10::intrusive_ptr<ivalue::PyObjectHolder> intrusive_from_this() {
```
this does not appear used

zdevito commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -390,6 +399,30 @@ struct C10_EXPORT ivalue::Object final : c10::intrusive_ptr_target {
+    return c10::intrusive_ptr<ivalue::PyObjectHolder>::reclaim(this);
+  }
+ public:
+  static c10::intrusive_ptr<ivalue::PyObjectHolder> create(PyObject* py_obj) {
```
This captures py_obj but does not call incref on it. It needs to have a name that suggests it is taking ownership of the py_obj, such as `PyObjectHolder::steal(py_obj)`

zdevito replied:
Actually, this probably cannot go here because refcounting logic needs to happen in libtorch_python

zdevito commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -538,6 +538,9 @@ inline IValue toIValue(
     }
+    case TypeKind::PyObjectType:
+      // convert a pybind11 object to the IValue and hold it
+      return c10::ivalue::PyObjectHolder::create(obj.ptr());
```
There is not incref here. If the python object goes dead then the reference in the by object holder is invalidated.

zdevito commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -717,6 +720,8 @@ inline py::object toPyObject(IValue ivalue) {
       py::setattr(pyObj, attrName.c_str(), toPyObject(std::move(v)));
     }
     return pyObj;
+  } else if (ivalue.isPyObject()) {
```
py::reinterpret_borrow, otherwise this will not incref the underlying pyobject

zdevito commented on test/test_jit.py:
```diff
@@ -2798,6 +2798,18 @@ def fn(x):
         fn(y)
 
+    def test_python_ivalue(self):
+        # Test if pure python object can be hold as IValue and conversion
```
This test should test memory management: if IValue is the only holder of the object then it should not go dead. It should also not cause the python object to leak.

zdevito reviewed (commented):
Inline comments have some suggestions about refcounting for the contained PyObjects.

wanchaol committed:
Update on "[jit] Enable IValue to hold a PyObject"


Summary:

This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 95e3bb6:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=95e3bb6a1313242db181722b5a34668143c6cf52).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3960762) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74887707))
<details>
<summary>
<code>Dec 16 08:58:25 /var/lib/jenkins/workspace/torch/csrc/jit/unpickler.cpp:73:12: error: enumeration value 'PyObjectType' not handled in switch [-Werror=switch]</code>
</summary>

```
Dec 16 08:58:12                  from /var/lib/jenkins/workspace/torch/csrc/jit/autodiff.h:4, 
Dec 16 08:58:12                  from /var/lib/jenkins/workspace/torch/csrc/jit/autodiff.cpp:1: 
Dec 16 08:58:12 /var/lib/jenkins/workspace/aten/src/ATen/core/ivalue_inl.h:687:8: note: declared here 
Dec 16 08:58:12  inline IValue::IValue(std::vector<int64_t> v) 
Dec 16 08:58:12         ^ 
Dec 16 08:58:12 [ 77%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/unpickler.cpp.o 
Dec 16 08:58:12 [ 77%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/graph_executor.cpp.o 
Dec 16 08:58:22 [ 77%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/import_source.cpp.o 
Dec 16 08:58:22 [ 77%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/import.cpp.o 
Dec 16 08:58:25 /var/lib/jenkins/workspace/torch/csrc/jit/unpickler.cpp: In function 'void torch::jit::restoreAccurateTypeTags(const c10::IValue&, const TypePtr&)': 
Dec 16 08:58:25 /var/lib/jenkins/workspace/torch/csrc/jit/unpickler.cpp:73:12: error: enumeration value 'PyObjectType' not handled in switch [-Werror=switch] 
Dec 16 08:58:25      switch (w.static_type->kind()) { 
Dec 16 08:58:25             ^ 
Dec 16 08:58:25 /var/lib/jenkins/workspace/torch/csrc/jit/unpickler.cpp: In instantiation of 'c10::IValue torch::jit::toSpecializedList(const c10::IValue&) [with T = long int]': 
Dec 16 08:58:25 /var/lib/jenkins/workspace/torch/csrc/jit/unpickler.cpp:467:70:   required from here 
Dec 16 08:58:25 /var/lib/jenkins/workspace/torch/csrc/jit/unpickler.cpp:238:39: warning: 'c10::IValue::IValue(std::vector<long int>)' is deprecated: IValues based on std::vector<T> are potentially slow and deprecated. Please use c10::List<T> instead. [-Wdeprecated-declarations] 
Dec 16 08:58:25    return IValue(std::move(specialized)); 
Dec 16 08:58:25                                        ^ 
Dec 16 08:58:25 In file included from /var/lib/jenkins/workspace/aten/src/ATen/core/ivalue.h:661:0, 
Dec 16 08:58:25                  from /var/lib/jenkins/workspace/aten/src/ATen/core/jit_type.h:6, 
Dec 16 08:58:25                  from /var/lib/jenkins/workspace/aten/src/ATen/core/function_schema.h:4, 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

wanchaol committed:
Update on "[jit] Enable IValue to hold a PyObject"


Summary:

This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Enable IValue to hold a PyObject"


Summary:

This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol commented on torch/csrc/jit/python_ivalue.h:
```diff
@@ -0,0 +1,37 @@
+#pragma once
```
I have to create this as header because pybind_utils.h need to use it for conversions.

wanchaol committed:
Update on "[jit] Enable IValue to hold a PyObject"


Summary:

This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -543,6 +544,10 @@ inline IValue toIValue(
+    case TypeKind::PyObjectType:
+      // convert a pybind11 object to the IValue that holds the
+      // PyObject and steal the reference to take ownership of it.
+      return c10::ivalue::ConcretePyObjectHolder::steal(obj.ptr());
```
This steals the reference meaning somewhere else in the code a decref needs to _not_ happen.  But I see no code that would accomplish this. Should this not borrow the reference (i.e incref the object before handing to PyObjectHolder)

zdevito commented on torch/csrc/jit/python_ivalue.h:
```diff
@@ -0,0 +1,37 @@
+// concrete ivalue Holder that hold a PyObject*
+struct C10_EXPORT ConcretePyObjectHolder final : PyObjectHolder {
+ public:
+  static c10::intrusive_ptr<PyObjectHolder> steal(PyObject* py_obj) {
```
Since this already include pybind11, this can just take and store py::object, which would remove all the need for any memory management. The problem before was that the holder was not abstract.

zdevito replied:
This isn't a  steal if the constructor is calling incref.

zdevito reviewed (commented):
See inline comments about memory management. steal does not reflect what is happening and concretePyobjectHolder can just hold the py::object rather than need manual memory management.

wanchaol committed:
Update on "[jit] Enable IValue to hold a PyObject"


Summary:

This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Enable IValue to hold a PyObject"


Summary:

This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Enable IValue to hold a PyObject"


Summary:

This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Enable IValue to hold a PyObject"


Summary:

This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/python_ivalue.h:
```diff
@@ -0,0 +1,31 @@
+    return py_obj_.ptr();
+  }
+
+  ~ConcretePyObjectHolder() {}
```
nit: it's not necessary to explicitly define the destructor.

zdevito reviewed (approved):
Looks good, thanks!

wanchaol committed:
Update on "[jit] Enable IValue to hold a PyObject"


Summary:

This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zhaojuanmao commented:
rebased and made minor lint changes, then exported as a new PR #32453 right now, so closing this one, as it is duplicated of #32453
